lib/class-wp-rest-url-details-controller.php

```diff
@@ -0,0 +1,246 @@
+		 * @param string $cache_key the generated cache key for the given URL.
+		 */
+		if ( apply_filters( 'rest_url_details_use_cache', $use_cache, $url, $cache_key ) && ! is_wp_error( $data ) && ! empty( $data ) ) {
+			return json_decode( $data, true );
+		}
+
+		$response_body = $this->get_remote_url( $url );
+
+		// Exit if we don't have a valid body or it's empty.
+		if ( is_wp_error( $response_body ) || empty( $response_body ) ) {
+			return $response_body;
+		}
+
+		$data = array(
+			'title' => $this->get_title( $response_body ),
+		);
+
+		// Cache the valid response.
+		$this->set_cache( $cache_key, $data );
+
+		return rest_ensure_response( $data );
+	}
+
+	/**
+	 * Checks whether a given request has permission to read remote urls.
+	 *
+	 * @return WP_Error|bool True if the request has access, WP_Error object otherwise.
+	 */
+	public function permissions_check() {
+		if ( current_user_can( 'edit_posts' ) ) {
+			return true;
+		}
+
+		foreach ( get_post_types( array( 'show_in_rest' => true ), 'objects' ) as $post_type ) {
+			if ( current_user_can( $post_type->cap->edit_posts ) ) {
+				return true;
+			}
+		}
+
+		return new WP_Error(
+			'rest_user_cannot_view',
+			__( 'Sorry, you are not allowed to process remote urls.', 'gutenberg' ),
+			array( 'status' => rest_authorization_required_code() )
+		);
+	}
+
+	/**
+	 * Retrieves the document title from a remote URL.
+	 *
+	 * @param string $url The website url whose HTML we want to access.
+	 * @return string|WP_Error The URL's document title on success, WP_Error on failure.
+	 */
+	private function get_remote_url( $url ) {
+
+		$args = array(
+			'timeout'             => 10,
+			'limit_response_size' => 153600, // 150 KB.
+		);
+
+		/**
+		 * Filters the HTTP request args for URL data retrieval.
+		 *
+		 * Can be used to adjust timeout and response size limit.
+		 *
+		 * @param array $args Arguments used for the HTTP request
+		 * @param string $url The attempted URL.
+		 */
+		$args = apply_filters( 'rest_url_details_http_request_args', $args, $url );
+
+		$response = wp_safe_remote_get(
+			$url,
+			$args
+		);
+
+		if ( WP_Http::OK !== wp_remote_retrieve_response_code( $response ) ) {
+			// Not saving the error response to cache since the error might be temporary.
+			return new WP_Error( 'no_response', get_status_header_desc( 404 ), array( 'status' => WP_Http::NOT_FOUND ) );
+		}
+
+		$remote_body = wp_remote_retrieve_body( $response );
+
+		if ( empty( $remote_body ) ) {
+			return new WP_Error( 'no_content', __( 'Unable to retrieve body from response at this URL.', 'gutenberg' ), array( 'status' => WP_Http::NOT_FOUND ) );
+		}
+
+		return $remote_body;
+	}
+
+	/**
+	 * Parses the <title> contents from the provided HTML
+	 *
+	 * @param string $html the HTML from the remote website at URL.
+	 * @return string the title tag contents (maybe empty).
+	 */
+	private function get_title( $html ) {
+		preg_match( '|<title>([^<]*?)</title>|is', $html, $match_title );
```

Comment:
Might need to allow for arbitrary attributes on the `<title>`?

Reply:
That sounds like a good plan. Here's a first pass:

```
<title( [a-z]+="[\w\d\s]+")*>([^<]*?)<\/title>
```

Ultimately we're probably going to use `DOMDocument` to better "query" the HTML, but I'd like to get this "right".

Reply:
It gets kind of painful trying to match all possible (valid, let alone "allowed") strings here, this SO post tries to get there -- https://stackoverflow.com/questions/317053/regular-expression-for-extracting-tag-attributes

Maybe we have something somewhere else in WP code that we can re-use?

> Ultimately we're probably going to use DOMDocument to better "query" the HTML, but I'd like to get this "right".
👍

Reply:
Maybe https://regex101.com/r/U9Yqqg/2.

The quicker we follow up with something more robust the better!

Reply:
Hi! What kind of extra title attributes would this use?

Reply:
It is an edge case for sure.

We could just roll with not supporting attrs given we know I will upgrade to a more robust method of interogating the HTML in a follow up.

Reply:
> Hi! What kind of extra title attributes would this use?

FWIW my suggestion was just to make sure that we allow for (and ignore) them, not that we actually need to use them for anything. I don't know that they are very common, but they are [technically valid](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/title#Attributes).

Reply:
> We could just roll with not supporting attrs given we know I will upgrade to a more robust method of interogating the HTML in a follow up.

Yes to this imho! Thanks @beaulebens for the clarification!

Reply:
Raised https://github.com/WordPress/gutenberg/issues/27762 to followup.